### PR TITLE
[Merged by Bors] - feat: `#trans_imports` shows the imports of the current module

### PR DESCRIPTION
**Pull request by adomani**
`#trans_imports` reports how many transitive imports the current module has.
The command takes two optional inputs:
* `#trans_imports str`, where `str` is a `String` also shows the transitively imported modules whose name begins with `str`;
* `#trans_imports (str)? at most n` behaves like `#trans_imports (str)?`, except that instead of reporting the *exact* count of transitive imports, it either confirms that their number is at most `n`, or warns that it is larger.

The `at_most n` option is useful for tests, but could potentially be used to limit import growth in mathlib.

[Zulip discussion](https://leanprover.zulipchat.com/#narrow/channel/287929-mathlib4/topic/import.20changes.20bot.20on.20PRs/near/482768569)

---

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

File: MathlibTest/TransImports.lean [hunk 1]
```diff
@@ -0,0 +1,10 @@
+import Mathlib.Util.TransImports
+
+/--
+info: 'MathlibTest.TransImports' has 483 transitive imports
```
Comment (adomani) [inline MathlibTest/TransImports.lean:4]: Having the tests fail when the actual *number* of transitive imports changes may be a little too volatile.

File: Mathlib/Util/TransImports.lean [hunk 2]
```diff
@@ -0,0 +1,31 @@
+/-
+Copyright (c) 2024 Damiano Testa. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Damiano Testa
+-/
+
+import Mathlib.Init
```
Comment (alreadydone) [inline Mathlib/Util/TransImports.lean:7]: Should this file be imported into Mathlib.Init instead? So that it's available in any Mathlib file.

Comment (adomani) [reply Mathlib/Util/TransImports.lean:7]: I imported it in `Mathlib.Tactic.Common`.  I think that there is the intention to make the imports of `Mathlib.Init` somewhat restricted : `Mathlib.Tactic.Common` is still imported in a large part of Mathlib anyway.

File: MathlibTest/TransImports.lean (hunk 1 above)
Comment (alreadydone) [reply MathlibTest/TransImports.lean:4]: Yeah ... should we leave out the test? Or (add an option to) remove the total count from the output?

**Review by alreadydone — APPROVED** [review]
Thanks! I'm not usually qualified to review meta PRs, but since I requested this PR, I asked Claude to [explain it](https://gist.github.com/alreadydone/4533d35def474651f235fdded6e1ae95) to me, and it looks good!

**Comment by alreadydone** [comment]
Do you know what metric the GitHub bot is using to report dependency/import changes? It seems to disagree with your script, e.g. it [reports](https://github.com/leanprover-community/mathlib4/pull/19527#issuecomment-2503060138) 1448 imports of Mathlib.FieldTheory.IsAlgClosed.Classification but your script says 1300 Mathlib imports / 2531 total. But I think your script does detect **changes* in imports accurately.

**Comment by adomani** [comment]
> Do you know what metric the GitHub bot is using to report dependency/import changes? It seems to disagree with your script, e.g. it [reports](https://github.com/leanprover-community/mathlib4/pull/19527#issuecomment-2503060138) 1448 imports of Mathlib.FieldTheory.IsAlgClosed.Classification but your script says 1300 Mathlib imports / 2531 total. But I think your script does detect *_changes_ in imports accurately.

The script used in the PR summary is a text-based script, so it may have some weirdness analysing files that `import [non-mathlib]` and also files that have
````lean
```example
import Mathlib
#check myNewTheorem
```
in them.  I do not remember the details of how that script works, though.

This may explain why relative diffs are more consistent, since I expect these effects to be detected only on the fringes of mathlib.
````

File: MathlibTest/TransImports.lean (hunk 1 above)
Comment (adomani) [reply MathlibTest/TransImports.lean]: I decided to add a second optional argument to the command: `#trans_imports "Lean.Elab.I" at_most 500` confirms that the number of transitive imports is at most 500, but does not print the actual number.  This should be more stable and might even be useful to limit the import size of certain files.

Comment (alreadydone) [reply MathlibTest/TransImports.lean]: Thanks!
maintainer merge

**Review by Vierkantor — APPROVED** [review]
Thanks!

I agree that filtering on module name is a good requirement for listing, otherwise we'd get thousands of irrelevant results each time we run the command.

bors merge

**Pull request closed without merging**